### DISCUSSION TITLE
Provide gradle idiomatic configuration

### DISCUSSION THREAD

**Discussion by triplem** [detekt/detekt#3483] [Ideas]
In gradle 6.8.x (and most probably before this already) gradle provides an "idiomatic" way to define aggregrations of reports without task-dependencies. This is mainly done using configurations, for an example see https://docs.gradle.org/release-nightly/samples/sample_structuring_software_projects.html.

## Expected Behavior
The detekt Task should support this idiomatic way and should (best case) provide some detailed documentation on how to apply this pattern to an existing project. 

## Current Behavior
The current plugin implementation does not provide a nice and simple way to implement an report-aggregation, without any task-dependencies without a rootProject-build.

**Comment by cortinico** [comment]
_Moving to a discussion_

We're already working on consolidating reports here: https://github.com/detekt/detekt/pull/3394 - @chao2zhang is on it and we will hopefully come up with a solution that works with Gradle in the near future 👍 

> gradle provides an "idiomatic" way to define aggregrations of reports without task-dependencies

I'd say that there is no "idiomatic" way to define aggregations. What they're doing is defining a precompiled gradle plugins + a new `configuration` used to feed the dependency to Jacoco. That probably makes sense for Jacoco as it needs to access the classpath. In our case it's sufficient to access the singular local reports and merge them.
I'd say that having a `detektReporting` configuration will probably create more confusion for our users than actual benefit (if compared to a Gradle plugin/task that takes care of dependencies and merging automatically).

**Comment by triplem** [comment]
This solution has the advantage, that no additional dependencies need to be declared, as well as the aggregation could be done in any project, not only the root Project. 

But I do unterstand your comment. Thanks.

**Comment by picklebento** [comment]
This is my first time reading "Structuring Large Projects" and the sample was using a particular set of gradle features such as composite builds. 
- This is actually very different from the large projects that I am working on while we are still using a mixture of standalone plugins and included script.
- I am also curious to learn if the same project is scalable that the build speed performance is optimized. In the projects that I am working on, dependencies and new configurations become very tricky as any small performance inefficiency would be magnified.

Hence in #3394 the consensus is to provide Gradle tasks and sample code to aggregate these reports.
Meanwhile, some tools have already supported multi-report aggregation natively, such as https://detekt.github.io/detekt/reporting.html#integration-with-github-code-scanning. The `sarif_file` can be a folder that the Github action will recursively look for SARIF reports under the folder. (Also because I added the support in https://github.com/github/codeql-action/pull/352)

**Reply by triplem** [reply]
If the produced report (SARIF, xml) is provided as an ouput of the subproject tasks, and as an input on the aggregation task, then an easy chaining of tasks is possible using the above mentioned method. If the aggregation task provides the aggregated report as an output, then this output can easily be used in eg. the sonarqube tasks (provided, that they do use the SourceTasks as the base).
This method allows, that the Sonarqube task acesses the output of the previous task and this task is called if any changes on their input occured.

I do like this solution, and think, that this is a rather clean solution and not far from the one, described in the mentioned links.

**Reply by BraisGabin** [reply]
The problem is that detekt, right now, have multiple tasks that you can run. For example in my project. It is an android project. Some modules have flavors, others not and some of them are pure jvm modules. It is impossible to chain inputs and outputs here because detekt doesn't know which task I want to run (and if I only want to run one of them, maybe I want to run all the flavors).

We need to find an easy way to solve this first. Then, probably, we can do your propouse. But meanwhile we can provide a script so anyone can configure the task as he/she wants.

**Reply by triplem** [reply]
Basically with the above solution the user can already decide by selecting the right Task by Type and name. And flavors are supported by design.

**Reply by BraisGabin** [reply]
But can't create a consolidated report mixing different types of detekt. That's why, right now, the scritpt is the best solution. It's not a final solution but we can't provide a solution that only works for a part of our users. It's counter intuitive but I prefer to not have the feature that have the feature that only works for some users (even if they are the 90%). But we have other ways to provide "non official" solutions for those 90% of users: blog posts, comments in the documentation, etc.
